### PR TITLE
perf: replace enums with dynamically derived object literals

### PR DESCRIPTION
**Pull request by 0xApotheosis**
Replaces `ChainNamespace`, `ChainReference`, `AssetNamespace`, `AssetReference` enums with types dynamically derived from object literals.

This gives us literal string union types, a single source of truth for each reference/namespace, and object literals for good developer experience. String literals will also type check correctly.

For example:

```ts
export type AssetNamespace = typeof ASSET_NAMESPACE[keyof typeof ASSET_NAMESPACE]

const test1: AssetNamespace = 'cw20' // ✅ passes type check
const test2: AssetNamespace = ASSET_NAMESPACE.CW20 // ✅ passes type check
const test3: AssetNamespace = 'invalid-string' // 🛑 fails type check
```

Closes https://github.com/shapeshift/lib/issues/550

Screen Shot 2022-05-17 at 7 16 29 pm

### CODE REVIEW

File: packages/caip/src/adapters/osmosis/utils.ts [hunk 1]
```diff
@@ -39,11 +39,11 @@ export const parseOsmosisData = (data: OsmosisCoin[]) => {
 
     if (isNativeAsset) {
       // TODO(ryankk): remove `toString` when AssetReferences are changed to strings
-      assetReference = isOsmo ? AssetReference.Osmosis.toString() : denom
-      assetNamespace = isOsmo ? AssetNamespace.Slip44 : AssetNamespace.NATIVE
+      assetReference = isOsmo ? ASSET_REFERENCE.Osmosis.toString() : denom
```
Comment (kaladinlight) [inline packages/caip/src/adapters/osmosis/utils.ts:42]: Can remove `toString()` and comment above

File: packages/investor-foxy/src/api/api.ts [hunk 2]
```diff
@@ -53,14 +53,14 @@ import {
 
 export * from './foxy-types'
 
+// EthereumMainnet, EthereumRinkeby, EthereumRopsten
+type Network = '1' | '3' | '4'
```
Comment (kaladinlight) [inline packages/investor-foxy/src/api/api.ts:57]: Reason to not use constant strings over hardcode?

**Review by elmutt — APPROVED** [review]

File: packages/caip/src/adapters/osmosis/utils.ts (hunk 1 above)
Comment (0xApotheosis) [reply packages/caip/src/adapters/osmosis/utils.ts]: Sounds good - removed in https://github.com/shapeshift/lib/pull/656/commits/f2b3ea985a4f48f4d3e7ba2e14bd3491cfecf48c

File: packages/investor-foxy/src/api/api.ts (hunk 2 above)
Comment (0xApotheosis) [reply packages/investor-foxy/src/api/api.ts:57]: There was an issue using the `CHAIN_REFERENCE` constant due to a circular import, or something similar.

It was a battle I chose not to fight last night - I'll take another look today and add an inline comment for clarity if I can't work it out.

Comment (kaladinlight) [reply packages/investor-foxy/src/api/api.ts:57]: Circular import strikes me as odd considering the caip package where they are defined should not be importing any "feature" packages in lib. But that sounds good, let me know.

Comment (0xApotheosis) [reply packages/investor-foxy/src/api/api.ts]: Committed what we landed on in https://github.com/shapeshift/lib/pull/656/commits/8226d6f9308f7999c4154eb4c3b36ae6be5c7d39

**Review by kaladinlight — APPROVED** [review]

**Comment by 0xApotheosis** [comment]
Don't merge yet - this should technically be a breaking change.

**Review by elmutt — APPROVED** [review]

**Comment by 0xApotheosis** [comment]
Converting to draft until I get a corresponding web PR ready.

**Pull request merged**